qiskit/mitigation/correlated_readout_mitigator.py

```diff
@@ -0,0 +1,231 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Readout mitigator class based on the A-matrix inversion method
+"""
+
+from typing import Optional, List, Tuple, Iterable, Callable, Union
+import numpy as np
+
+from qiskit.result import Counts, QuasiDistribution
+from .base_readout_mitigator import BaseReadoutMitigator
+from .utils import counts_probability_vector, z_diagonal, str2diag
+
+
+class CorrelatedReadoutMitigator(BaseReadoutMitigator):
+    """N-qubit readout error mitigator.
+    Mitigates expectation_value and quasi_probabilities.
+    The mitigation_matrix should be calibrated using qiskit.experiments."""
+
+    def __init__(self, amat: np.ndarray, qubits: Optional[Iterable[int]] = None):
+        """Initialize a CorrelatedReadoutMitigator
+        Args:
+            amat: readout error assignment matrix.
+            qubits: Optional, the measured physical qubits for mitigation.
+        """
+        if qubits is None:
+            self._num_qubits = int(np.log2(amat.shape[0]))
+            self._qubits = range(self._num_qubits)
+        else:
```

Comment:
For this case it would be good to validate the number of qubits matches the matrix and raise an exception if it does not.

Reply:
Done